src/seals/diagnostics/base_env.py

```diff
@@ -0,0 +1,100 @@
+"""Minimal base class for simple environments."""
+
+import gym
+from gym.spaces import Discrete
+from gym.utils import seeding
+import numpy as np
+
+from seals.util import sample_distribution
+
+
+class BaseEnv(gym.Env):
+    """Minimal environment class, with default tabular method implementations."""
+
+    def __init__(self, num_states=None, num_actions=None):
+        """Initialize spaces (if discrete) and select random seed."""
+        super().__init__()
+
+        if num_states is not None:
```

Comment:
This is confusing: caller doesn't have to define `observation_space` if they specify this but otherwise they do.

Better to make sure mandatory properties are always set. If you split this up into a tabular and non-tabular part it'd be natural; have the non-tabular part take mandatory `observation_space` and `action_space` arguments, and the tabular part take mandatory `num_states` and `num_action` arguments.